src/lib/blkalloc/varsize_blk_allocator.cpp

```diff
@@ -405,251 +384,380 @@ void VarsizeBlkAllocator::fill_cache_in_portion(blk_num_t portion_num, blk_cache
-                         excess_blks.size());
-            free_on_bitmap(blk_cache_entry_to_blkid(e));
+    blk_count_t n_freed{0};
+    if (bid.is_multi()) {
+        auto it = bid.iterate();
+        while (auto const b = it.next()) {
+            n_freed += do_free(*b);
         }
     } else {
-        // free directly on bitmap
-        free_on_bitmap(b);
+        n_freed += do_free(bid);
     }
 
-    decr_alloced_blk_count(b.get_nblks());
-    BLKALLOC_LOG(TRACE, "Freed blk_num={}", blkid_to_blk_cache_entry(b).to_string());
+    for (auto const& e : excess_blks) {
+        BLKALLOC_LOG(TRACE, "Freeing in bitmap of entry={} - excess of free_blks size={}", e.to_string(),
+                     excess_blks.size());
+        free_blks_direct(MultiBlkId{blk_cache_entry_to_blkid(e)});
+    }
+    return n_freed;
 }
 
-blk_cap_t VarsizeBlkAllocator::available_blks() const { return get_total_blks() - get_used_blks(); }
-blk_cap_t VarsizeBlkAllocator::get_used_blks() const { return get_alloced_blk_count(); }
+blk_count_t VarsizeBlkAllocator::free_blks_direct(MultiBlkId const& bid) {
+    auto const do_free = [this](BlkId const& b) {
+        BlkAllocPortion& portion = blknum_to_portion(b.blk_num());
+        {
+            auto const start_blk_id = portion.get_portion_num() * get_blks_per_portion();
+            auto const end_blk_id = start_blk_id + get_blks_per_portion() - 1;
+            auto lock{portion.portion_auto_lock()};
+            HS_DBG_ASSERT_LE(start_blk_id, b.blk_num(), "Expected start bit to be greater than portion start bit");
+            HS_DBG_ASSERT_GE(end_blk_id, (b.blk_num() + b.blk_count() - 1),
+                             "Expected end bit to be smaller than portion end bit");
+            BLKALLOC_REL_ASSERT(m_cache_bm->is_bits_set(b.blk_num(), b.blk_count()), "Expected bits to be set");
+            m_cache_bm->reset_bits(b.blk_num(), b.blk_count());
+            portion.increase_available_blocks(b.blk_count());
+        }
+        BLKALLOC_LOG(TRACE, "Freeing directly to portion={} blkid={} set_bits_count={}",
+                     blknum_to_portion_num(b.blk_num()), b.to_string(), get_alloced_blk_count());
+        return b.blk_count();
+    };
+
+    blk_count_t n_freed{0};
+    if (bid.is_multi()) {
+        auto it = bid.iterate();
+        while (auto const b = it.next()) {
+            n_freed += do_free(*b);
+        }
+    } else {
+        n_freed += do_free(bid);
+    }
+    return n_freed;
+}
 
-void VarsizeBlkAllocator::free_on_bitmap(const BlkId& b) {
-    BlkAllocPortion& portion = blknum_to_portion(b.get_blk_num());
-    {
-        auto const start_blk_id = portion.get_portion_num() * get_blks_per_portion();
-        auto const end_blk_id = start_blk_id + get_blks_per_portion() - 1;
+bool VarsizeBlkAllocator::is_blk_alloced(BlkId const& b, bool use_lock) const {
+    if (!m_inited) { return true; }
```

Comment:
do we need to handle MultiBlkId here?

Reply:
I will add that code, but as is is_blk_alloced is not used or this code is to be modified not to assert here.